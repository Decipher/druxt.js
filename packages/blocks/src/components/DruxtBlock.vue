<script>
import { DrupalJsonApiParams } from 'drupal-jsonapi-params'
import DruxtModule from 'druxt/dist/components/DruxtModule.vue'
import { mapActions } from 'vuex'

/**
 * The DruxtBlock component is used to render a Drupal Block by UUID or Drupal's
 * internal ID.
 *
 * `<DruxtBlock id="umami_branding" />`
 *
 * `<DruxtBlock uuid="59104acd-88e1-43c3-bd5f-35800f206394" />`
 *
 * While the DruxtBlock component can't automatically render every Drupal block,
 * it does provide the Block settings to a targetted Druxt wrapper component for
 * manual theming.
 *
 * @example @lang vue
 * <DruxtBlock id="umami_branding" />
 *
 * @example @lang vue
 * <DruxtBlock uuid="59104acd-88e1-43c3-bd5f-35800f206394" />
 */
export default {
  name: 'DruxtBlock',

  extends: DruxtModule,

  /**
   * The DruxtBlock component requires either the UUID or internal ID property.
   */
  props: {
    /**
     * The Blocks internal ID.
     *
     * @type string
     *
     * @example @lang vue
     * <DruxtBlock id="umami_branding" />
     */
    id: {
      type: String,
      default: null,
    },

    /**
     * The Block Entity UUID.
     *
     * @type {string}
     *
     * @example @lang vue
     * <DruxtBlock uuid="59104acd-88e1-43c3-bd5f-35800f206394" />
     */
    uuid: {
      type: String,
      default: null,
    },
  },

  /**
   * @property {object} resource - The JSON:API resource object.
   */
  data: () => ({
    resource: {},
  }),

  fetchKey(getCounter) {
    const parts = ['DruxtBlock', this.uuid || this.id].filter((o) => o)
    return [...parts, getCounter(parts.join(':'))].join(':')
  },

  /** */
  computed: {
    /**
     * The Block entity data.
     *
     * @return {object}
     */
    block: ({ resource }) => (resource || {}).data,
  },

  methods: {
    /**
     * Maps Vuex action to methods.
     */
    ...mapActions({
      getCollection: 'druxt/getCollection',
      getResource: 'druxt/getResource',
    })
  },

  /**
   * Druxt module configuration.
   */
  druxt: {
    /**
     * Provides the available component naming options for the DruxtWrapper.
     *
     * @param {object} context - The module component ViewModel.
     * @returns {ComponentOptions}
     */
    componentOptions: ({ block }) => {
      if (!block) return []

      // Get Plugin and Plugin ID data.
      let plugin = block.attributes.plugin || ''
      let pluginId = null
      if (plugin.includes(':')) {
        [plugin, pluginId] = plugin.split(':')
      }

      // Construct component options.
      const componentOptions = []
      if (pluginId) {
        componentOptions.push([plugin, pluginId, block.attributes.region, block.attributes.theme])
        componentOptions.push([plugin, pluginId, block.attributes.theme])
      }
      componentOptions.push([plugin, block.attributes.region, block.attributes.theme])
      componentOptions.push([plugin, block.attributes.theme])
      componentOptions.push(['default'])
      return componentOptions
    },

    /**
     * Fetches the Block JSON:API resource by either UUID or ID.
     */
    async fetchConfig() {
      // Build query.
      const type = 'block--block'
      const query = new DrupalJsonApiParams()

      const fields = ((this.$druxt.settings.blocks || {}).query || {}).fields
      if (Array.isArray(fields)) {
        query.addFields(type, [
          ...fields,
          'plugin',
          'region',
          'settings',
          'theme',
        ])
      }

      // Fetch Block by UUID.
      if (this.uuid) {
        const id = this.uuid
        this.resource = await this.getResource({ type, id, query })
      }

      // Fetch Block by Drupal internal ID.
      else if (this.id) {
        query.addFilter('drupal_internal__id', this.id)
        const collection = await this.getCollection({ type, query })
        this.resource = { data: collection.data[0] }
      }
    },

    /**
     * Provides propsData for the DruxtWrapper.
     *
     * @param {object} context - The module component ViewModel.
     * @returns {PropsData}
     */
    propsData: ({ block }) => ({ block }),

    /**
     * Provides the scoped slots object for the Module render function.
     *
     * A default slot is provided with debug information if Nuxt is in
     * development mode.
     *
     * @return {ScopedSlots} The Scoped slots object.
     */
    slots(h) {
      const scopedSlots = {}

      // Debug data.
      /* @slot Debug information */
      scopedSlots.default = () => {
        let summary = `Placeholder for the '${((this.block || {}).attributes || {}).drupal_internal__id}' block.`
        let description = [
          h('p', 'DruxtBlocks knows that a block can be rendered, and has information provided by Drupal, but not enough to automatically determine the behaviour of the block.'),
          h('p', 'To render this block manually, create a Nuxt component with one of the following component options.'),
        ]

        // Ensure an ID or UUID.
        if (!this.id && !this.uuid) {
          summary = "Missing required 'id' or 'uuid' prop."
          description = [h('p', "The DruxtBlock component requires either the 'id' or 'uuid' prop to be set.")]
        }

        return h('DruxtDebug',
          { props: { summary } },
          [
            h('div', description),
            !!this.component.options.length && h('label', ['Component options:', h('ul', this.component.options.map((s) => h('li', [s])))]),
            ((this.block || {}).attributes || {}).settings && h('label', ['Block settings:', h('pre', [JSON.stringify(this.block.attributes.settings, null, '  ')])])
          ]
        )
      }

      return scopedSlots
    },
  },
}

/**
 * Provides the available naming options for the Wrapper component.
 *
 * @typedef {array[]} ComponentOptions
 *
 * @example @lang js
 * [
 *   'DruxtBlock[Plugin][PluginId][Region][Theme]',
 *   'DruxtBlock[Plugin][PluginId][Theme]',
 *   'DruxtBlock[Plugin][Region][Theme]',
 *   'DruxtBlock[Plugin][Theme]',
 *   'DruxtBlock[Default]',
 * ]
 *
 * @example <caption>Umami Branding block</caption> @lang js
 * [
 *   'DruxtBlockSystemBrandingBlockHeaderUmami',
 *   'DruxtBlockSystemBrandingBlockHeader',
 *   'DruxtBlockSystemBrandingBlockUmami',
 *   'DruxtBlockSystemBrandingBlock',
 *   'DruxtBlockDefault',
 * ]
 */

/**
 * Provides settings for the Blocks module, via the `nuxt.config.js`
 * `druxt.blocks` object.
 *
 * @typedef {object} ModuleSettings
 * @param {string[]} fields - An array of fields to filter from the JSON:API Resource.
 *
 * @example @lang js
 * {
 *   fields: [],
 * }
 *
 * @example <caption>nuxt.config.js</caption> @lang js
 * export default {
 *   druxt: {
 *     blocks: {
 *       query: {
 *         fields: ['dependencies'],
 *       },
 *     },
 *   },
 * }
 */

/**
 * Provides propsData for use in the Wrapper component.
 *
 * @typedef {object} PropsData
 * @param {object} block - The Block Entity data.
 *
 * @example @lang js
 * {
 *   block: {
 *     attributes: {},
 *     id: '59104acd-88e1-43c3-bd5f-35800f206394',
 *     links: {},
 *     relationships: {},
 *     type: 'block--block',
 *   },
 * }
 */

/**
 * Provides scoped slots for use in the Wrapper component.
 *
 * @typedef {object} ScopedSlots
 * @param {function} default - Debug information if Nuxt is in development mode.
 */
</script>
