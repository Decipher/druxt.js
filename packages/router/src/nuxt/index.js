import { existsSync } from 'fs'
import { join, resolve } from 'path'

/**
 * The Nuxt.js module function.
 *
 * - Extends the Vue router, adding the Druxt wildcard route.
 * - Adds the Druxt router plugin.
 * - Adds the Druxt router Vuex store.
 *
 * The module function should not be used directly, but rather installed via your Nuxt configuration file.
 *
 * Options are set on the root level `druxt` Nuxt config object.
 *
 * @example @lang js
 * // `nuxt.config.js`
 * module.exports = {
 *   buildModules: ['druxt-router/nuxt'],
 *   druxt: {
 *     baseUrl: 'https://demi-api.druxtjs.org'
 *   }
 * }
 *
 * @property {object} options.druxt - Druxt root level options.
 * @property {string} options.druxt.baseUrl - Base URL of Drupal JSON:API backend.
 * @property {string} options.druxt.router.component - File to custom Router component.
 */
const DruxtRouterNuxtModule = async function (moduleOptions = {}) {
  // Set default options.
  const options = {
    baseUrl: moduleOptions.baseUrl,
    ...(this.options || {}).druxt || {},
    router: {
      pages: (await existsSync(resolve(this.options.srcDir, this.options.dir.pages))),
      wildcard: true,
      ...((this.options || {}).druxt || {}).router,
      ...moduleOptions,
    }
  }

  // Register components directories.
  this.nuxt.hook('components:dirs', dirs => {
    dirs.push({ path: join(__dirname, '../dist/components') })
  })

  // Add Druxt router custom wildcard route.
  if (options.router.wildcard) {
    this.addTemplate({
      src: resolve(__dirname, '../templates/component.js'),
      fileName: 'components/druxt-router.js',
      options
    })

    this.extendRoutes((routes, resolve) => {
      if (this.nuxt.options.i18n && this.nuxt.options.i18n.locales) {
        this.nuxt.options.i18n.locales.forEach((locale) => {
          routes.push({
            name: `druxt-router___${locale.code}`,
            path: `/${locale.code}/*`,
            component: options.router.component,
            chunkName: 'druxt-router',
            meta: { langcode: locale.code }
          })
        })
        routes.push({
          name: 'druxt-router',
          path: '/*',
          component: options.router.component,
          chunkName: 'druxt-router'
        })
      } else {
        // Add Druxt router custom wildcard route.
        routes.push({
          name: 'druxt-router',
          path: '*',
          component: options.router.component,
          chunkName: 'druxt-router'
        })
      }
    })
  }

  // Ignore page routes.
  if (options.router.wildcard && !options.router.pages) {
    this.nuxt.hook('build:before', () =>
      this.nuxt.options.build.createRoutes = () => []
    )
  }

  // Add plugin.
  this.addPlugin({
    src: resolve(__dirname, '../templates/plugin.js'),
    fileName: 'druxt-router.js',
    options
  })

  // Enable Vuex Store.
  this.options.store = true

  // Add Vuex plugin.
  this.addPlugin({
    src: resolve(__dirname, '../templates/store.js'),
    fileName: 'store/druxt-router.js',
    options
  })

  // Nuxt Storybook.
  this.nuxt.hook('storybook:config', async ({ stories }) => {
    stories.push('druxt-router/dist/components/*.stories.mjs')
  })
}

DruxtRouterNuxtModule.meta = require('../package.json')

export default DruxtRouterNuxtModule
