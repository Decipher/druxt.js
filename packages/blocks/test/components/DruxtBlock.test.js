import 'regenerator-runtime/runtime'
import { createLocalVue, mount } from '@vue/test-utils'
import Vuex from 'vuex'
import mockAxios from 'jest-mock-axios'

import { DruxtClient, DruxtStore } from '../../../druxt/src'
import DruxtBlock from '../../src/components/DruxtBlock.vue'
import { getMockResource } from '../../../test-utils/src'

// Setup local vue instance.
const localVue = createLocalVue()
localVue.use(Vuex)

let store

const mountComponent = (propsData = {}, options = {}, mocks = {}) => {
  mocks = {
    $druxt: {
      settings: {
        blocks: {
          query: {
            fields: []
          }
        }
      }
    },
    $fetchState: { pending: false },
    $nuxt: {
      context: {
        isDev: false,
      }
    },
    ...mocks
  }
  return mount(DruxtBlock, { localVue, mocks, propsData, store, stubs: ['DruxtDebug'], ...options })
}

describe('Component - DruxtBlock', () => {
  beforeEach(() => {
    mockAxios.reset()

    // Setup vuex store.
    store = new Vuex.Store()

    DruxtStore({ store })
    store.$druxt = new DruxtClient('https://demo-api.druxtjs.org')

    store.app = { context: { error: jest.fn() }, store }
  })

  test('uuid', async () => {
    const mockPage = await getMockResource('block--block')
    const wrapper = mountComponent({ uuid: mockPage.data.id })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    // Fetch key.
    expect(DruxtBlock.fetchKey.call(wrapper.vm, jest.fn(() => 0))).toBe(`DruxtBlock:${mockPage.data.id}:0`)

    // Props
    expect(wrapper.vm.id).toBe(null)
    expect(wrapper.vm.uuid).toBe(mockPage.data.id)

    // DruxtModule
    expect(wrapper.vm.component.is).toBe('DruxtWrapper')

    // Default slot.
    expect(wrapper.vm.getScopedSlots().default).toStrictEqual(expect.any(Function))
  })

  test('uuid - pluginId', async () => {
    // TODO : Update test to use getMockResource instead of a hardcoded UUID.
    const wrapper = mountComponent({ uuid: '06251689-406e-4dc4-aab1-5fcf0e5f9ecb' })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    // Fetch key.
    expect(DruxtBlock.fetchKey.call(wrapper.vm, jest.fn(() => 0))).toBe('DruxtBlock:06251689-406e-4dc4-aab1-5fcf0e5f9ecb:0')

    // DruxtModule
    expect(wrapper.vm.component.options.length).toBe(9)
    expect(wrapper.vm.component.options).toStrictEqual([
      'DruxtBlockBlockContent9aadf4a1Ded64017A10dA5e043396edfBannerTopUmami',
      'DruxtBlockBlockContent9aadf4a1Ded64017A10dA5e043396edfBannerTop',
      'DruxtBlockBlockContent9aadf4a1Ded64017A10dA5e043396edfUmami',
      'DruxtBlockBlockContentBannerTopUmami',
      'DruxtBlockBlockContent9aadf4a1Ded64017A10dA5e043396edf',
      'DruxtBlockBlockContentBannerTop',
      'DruxtBlockBlockContentUmami',
      'DruxtBlockBlockContent',
      'DruxtBlockDefault',
    ])
  })

  test('id', async () => {
    const wrapper = mountComponent({ id: 'umami_branding' })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    // Fetch key.
    expect(DruxtBlock.fetchKey.call(wrapper.vm, jest.fn(() => 0))).toBe('DruxtBlock:umami_branding:0')

    // Props
    expect(wrapper.vm.id).toBe('umami_branding')
    expect(wrapper.vm.uuid).toBe(null)

    // DruxtModule
    expect(wrapper.vm.component.options).toStrictEqual([
      'DruxtBlockSystemBrandingBlockHeaderUmami',
      'DruxtBlockSystemBrandingBlockHeader',
      'DruxtBlockSystemBrandingBlockUmami',
      'DruxtBlockSystemBrandingBlock',
      'DruxtBlockDefault',
    ])

    expect(wrapper.vm.component.is).toBe('DruxtWrapper')
  })

  test('dev mode slot', async () => {
    const mocks = {
      $nuxt: {
        context: {
          isDev: true,
        },
      },
    }
    const wrapper = mountComponent({ uuid: '9d3d3a23-69f5-4c2d-9a00-287492a52987' }, {}, mocks)
    await wrapper.vm.$options.fetch.call(wrapper.vm)


    // Fetch key.
    expect(DruxtBlock.fetchKey.call(wrapper.vm, jest.fn(() => 0))).toBe('DruxtBlock:9d3d3a23-69f5-4c2d-9a00-287492a52987:0')

    // Default slot.
    const slot = wrapper.vm.getScopedSlots().default()
    expect(slot.tag.endsWith('DruxtDebug')).toBe(true)
  })

  test('custom default slot', async () => {
    const scopedSlots = { default: jest.fn() }
    const wrapper = mountComponent({ uuid: '06251689-406e-4dc4-aab1-5fcf0e5f9ecb' }, { scopedSlots })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    // Fetch key.
    expect(DruxtBlock.fetchKey.call(wrapper.vm, jest.fn(() => 0))).toBe('DruxtBlock:06251689-406e-4dc4-aab1-5fcf0e5f9ecb:0')

    wrapper.vm.getScopedSlots().default.call()
    expect(scopedSlots.default).toHaveBeenCalledWith({
      block: wrapper.vm.block,
    })
  })

  test('error - missing required props', () => {
    const wrapper = mountComponent({})
    // Ensure that error message matches snapshot.
    expect(wrapper.html()).toMatchSnapshot()
  })
})
