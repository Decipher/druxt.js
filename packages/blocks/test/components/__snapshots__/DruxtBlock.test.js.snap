// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - DruxtBlock error - missing required props 1`] = `
"<div>
  <div>
    <druxtdebug-stub>
      <div>
        <p>The DruxtBlock component requires either the 'id' or 'uuid' prop to be set.</p>
      </div>
    </druxtdebug-stub>
  </div>
</div>"
`;
